posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,286 @@
+  override def remove(name: ItemName): F[Unit] = F.delay(println(s"Removing item: $item"))
+}
+```
+
+Voilà! We got our streaming implementation of `findAll`.
+
+### Test interpreter
+
+That's all we wanted, but what about testing it? Sure, we might prefer to have a simple implementation by just using a plain `List`, so what can we possibly do?
+
+```scala
+object MemRepository extends ItemRepository[Id, List] {
+  private val mem = MutableMap.empty[String, Item]
+
+  override def findAll: List[Item] = mem.headOption.map(_._2).toList
+  override def find(name: ItemName): Id[Option[Item]] = mem.get(name.value)
+  override def save(item: Item): Id[Unit] = mem.update(item.name.value, item)
+  override def remove(name: ItemName): Id[Unit] = {
+    mem.remove(name.value)
+    ()
+  }
+}
+```
+
+That's pretty much it! We managed to abstract over the return type of `findAll` by just adding an extra parameter to our algebra.
+
+### About composition
+
+At this point the avid reader might have thought, what if I want to write a generic function that takes all the items (using `findAll`), applies some discounts and writes them back to the DB (using `save`)?
+
+Short answer is, you might want to define a different algebra where `findAll` and `save` have the same types (eg: both of them are streams) but in case you find yourself wanting to make this work with the current types then let's try and find out!
+
+```scala
+class DiscountProcessor[F[_], G[_]: Functor](repo: ItemRepository[F, G]) {
+
+  def process(discount: Double): F[Unit] = {
+    val rs: G[Item] = repo.findAll.map(item => item.copy(price = item.price * (1 - discount)))
+    // And now what? repo.save returns F[Unit]
+  }
+
+}
+```
+
+Here we have a problem. We managed to apply the discount to all the items in a generic way by having a `Functor` constraint and got back a type `G[Item]`. But now we don't have a generic way to go through all the items and save each of them in the DB. We could be thinking about `Traverse` but as I mentioned before, this abstraction can't represent a stream of items that could be potentially infinite.
+
+But if we really want to do this, there's something we can do that involves type lambdas and natural transformation (a.k.a. `FunctionK` in the Cats library), which is really out of the scope of this blog post so I won't dive into details:
+
+
+```scala
+class DiscountProcessor[F[_], G[_], H[_]](repo: ItemRepository[F, G])
+                                         (implicit F: Monad[F],
+                                                   G: Functor[G],
+                                                   H: Traverse[H],
+                                                   fk: FunctionK[G, λ[x => F[H[x]]]]) {
+
+  def process(discount: Double): F[Unit] = {
+    val items: G[Item] = repo.findAll.map(item => item.copy(price = item.price * (1 - discount)))
+    val fa: F[H[Item]] = fk.apply(items)
+    fa.flatMap(t => t.traverse(item => repo.save(item)).void)
+  }
+
+}
+
```

Comment:
I'm wondering if something simpler like
```scala
class DiscountProcessor[F[_], G[_]: Functor](repo: ItemRepository[F, G], join: G[F[Unit]] => F[Unit]) {

  def process(discount: Double): F[Unit] = {
    val items: G[Item] = repo.findAll.map(item => item.copy(price = item.price * (1 - discount)))
    val saved: G[F[Unit]] = items.map(repo.save)
    join(saved)
  }
}
```
would work for `StreamingDiscountInterpreter` and `ListDiscountInterpreter` without requiring to load all items in memory in `StreamingDiscountInterpreter`.

Reply:
I'm curious, how would you define `join`?

Reply:
For `StreamingDiscountInterpreter` its signature would be `Stream[IO, IO[Unit]] => IO[Unit]`, right? So `_.evalMap(identity).compile` should fit the types.

Reply:
You're completely right and have destroyed that section :smile: 

I need to put a bit more of thought on it to see how other situations could be affected, but from a first sight it looks very correct, will be updating it soon, thanks!

Reply:
Updated it @fthomas , thanks again :+1: